c10/core/TensorImpl.cpp

```diff
@@ -18,11 +18,19 @@ C10_DEFINE_int64(
 namespace c10 {
 
 at::Tensor& TensorImpl::grad() {
-  AT_ERROR("grad is not implemented for Tensor");
+  if (autograd_meta()) {
+    return autograd_meta()->grad();
+  } else {
+    AT_ERROR("grad is not implemented for Tensor");
```

Comment:
This error message is not too accurate now, right?

Reply:
I feel that it's still correct, because these functions still only work on Variable and not Tensor. A more detailed version might be `Cannot call grad on Tensor that doesn't contain autograd metadata`, but only Variable contains autograd metadata right now, so the original error message is probably ok

Reply:
I feel that it might be worthwhile to get rid of `set_requires_grad` / `requires_grad` / `grad` in TensorImpl and move them to Variable class, since people can only call those functions on a Variable anyway.